programs/bench.h

```diff
@@ -19,46 +19,140 @@ extern "C" {
 #define ZSTD_STATIC_LINKING_ONLY   /* ZSTD_compressionParameters */
 #include "zstd.h"     /* ZSTD_compressionParameters */
 
+typedef enum {
+    BMK_timeMode = 0,
+    BMK_iterMode = 1
+} BMK_loopMode_t;
+
+typedef enum {
+    BMK_both = 0,
+    BMK_decodeOnly = 1,
+    BMK_compressOnly = 2
+} BMK_mode_t;
+
+#define ERROR_STRUCT(baseType, typeName) typedef struct { \
+    int error;       \
+    baseType result; \
+} typeName
+
 typedef struct {
     size_t cSize;
     double cSpeed;   /* bytes / sec */
     double dSpeed;
 } BMK_result_t;
```

Comment:
This is sort of what `BMK_customResult_t` is, and `benchMem()` does use it as you described. I think the differences are .time (now .nanoSecPerRun) is just the inverse of nbRunsPerSec, and that customBench adds all returns of the function run on each `srcBuffers[i]` (so it assumes returns are the size written to in the destination buffer or an error code using `ZSTD_isError()`), which wouldn't make sense with other kinds of return values. 

I'm not quite sure how to best generalize this though, since having the an array field `size_t* returnValue` in the struct would need freeing, which was the same problem had by BMK_resultSet_t (making the caller free the result). The other option I've considered is for `benchCustom` to only accept a single source (so it could return just the return value of the passed in function), which also has some complications, as the only ways I've thought of for this to happen while maintaining the same functionality for `benchMem` and `fullbench` are calling `benchCustom` multiple times (once per block) or leaving the block division up to `*benchFn`. 

In the former solution, we would still need to measure initFn only once, so the first call would have the normal initFn and for later sources we would just have initFn do nothing. It would also complicate timeMode for multiple sources, since the logic for that is internal to customBench. Even with the changes, it seems like this implementation would have different cache behavior than the old benchmarking functions.

The latter seems unnatural since a lot of the benchmarking functions in fullbench don't seem to have a parameter for this (as far as I'm aware of), so it would need to be passed in and implemented. This would also only be possible in the case where the data is stored contiguously, or we'd have to use the former again.

`BMK_mode_t` was intended to be used with paramgrill (decode_only was already an existing flag, and compress_only could be useful for saving an unnecessary operation at times).